drivers/SmartThings/harman-luxury/src/api/nsdk.lua

```diff
@@ -0,0 +1,286 @@
+local cosock = require "cosock"
+local http = cosock.asyncify "socket.http"
+local ltn12 = require "ltn12"
+local url = require "net.url"
+local json = require "st.json"
+local net_utils = require "st.net_utils"
+local st_utils = require "st.utils"
+local log = require "log"
+
+----------------------------------------------------------
+--- Definitions
+----------------------------------------------------------
+
+local HTTP = "http://"
+local GET_DATA = "/api/getData"
+local GET_ROWS = "/api/getRows"
+local SET_DATA = "/api/setData"
+
+----------------------------------------------------------
+--- Functions
+----------------------------------------------------------
+
+local NSDK = {}
+
+--- Helpers ----------------------------------------------
+
+--- formats a list of roles into a formatted string suitable for the HTTP API
+---@param rolesList table<string>|nil
+---@return string
+local function format_roles(rolesList)
+  if rolesList == nil then
+    return "value"
+  else
+    local t = {}
+    for _, role in rolesList do
+      if role == "all" then
+        return "@all"
+      else
+        table.insert(t, role)
+      end
+    end
+    return table.concat(t, ",")
+  end
+end
+
+--- creates a JSON error mesage from a string
+---@class ErrMsg
+---@field error table<string>
+---@param msg string
+---@return ErrMsg
+local function error_msg(msg)
+  return {
+    error = {
+      message = msg
+    }
+  }
+end
+
+--- send HTTP request
+---@param u string
+---@return table, boolean, integer
+local function sendRequest(u)
+  local sink = {}
+  local result, code, _ = http.request {
+    url = u,
+    method = "GET",
+    sink = ltn12.sink.table(sink)
+  }
+  return sink, result, code
+end
+
+--- handle HTTP request reply according to returned code
+---@param func_name string
+---@param u string
+---@param sink table
+---@param result boolean
+---@param code integer
+---@param valLocationFunc function
+---@return boolean, boolean|number|string|table|ErrMsg
+local function handleReply(func_name, u, sink, result, code, valLocationFunc)
+  if code == 200 then -- OK
+    local ret, val = pcall(json.decode, table.concat(sink, ""))
+    val = valLocationFunc(val)
+    log.debug(string.format("Nsdk %s: received value:%s", func_name, st_utils.stringify_table(val)))
+    return ret, val
+  elseif code == 500 then -- ERROR
+    local ret, err = pcall(json.decode, sink[result])
```

Comment:
Same here, and everywhere else `pcall` is used; I would prefer to see `pcall` "catches" handled inline instead of being bubbled up.

Just to reiterate: I think it's okay that these functions are fallible. But I would like to see internal error handling with a relevant error based on the `pcall` failure reported in the returns, instead of just bubbling up the boolean transparently.